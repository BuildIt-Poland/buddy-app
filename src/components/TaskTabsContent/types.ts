import { Task } from 'buddy-app-schema';
import { CheckboxProps } from '../TaskCheckbox';

export interface TaskTabsContentProps extends CheckboxProps {
  tasks: Task[];
  uncompletedCount?: number;
  completedCount?: number;
}

export interface TransformedTasks {
  uncompletedTasks: Task[];
  completedTasks: Task[];
}
