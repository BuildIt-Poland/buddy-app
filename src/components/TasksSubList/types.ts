import { Task } from 'buddy-app-schema';
import { CheckboxProps } from '../TaskCheckbox';

export interface TasksSubListProps extends CheckboxProps {
  tasks: Task[];
  tabIndex: number;
  title?: string;
}
