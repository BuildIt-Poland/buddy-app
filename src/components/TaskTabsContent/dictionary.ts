interface TaskTabsContentDictionary {
  NO_TASKS_TITLE: string;
  NO_TASKS_BUDDY_SUBTITLE: string;
  NO_TASKS_NEWBIE_SUBTITLE: string;
  TITLE_UNCOMPLETED: string;
  TITLE_COMPLETED: string;
}

const TASK_TABS_CONTENT_DICTIONARY: TaskTabsContentDictionary = {
  NO_TASKS_TITLE: 'No Tasks Found',
  NO_TASKS_BUDDY_SUBTITLE: 'You can always add a new task',
  NO_TASKS_NEWBIE_SUBTITLE: 'Ask your buddy to add a new task',
  TITLE_UNCOMPLETED: 'Todo',
  TITLE_COMPLETED: 'Done',
};

export default TASK_TABS_CONTENT_DICTIONARY;
