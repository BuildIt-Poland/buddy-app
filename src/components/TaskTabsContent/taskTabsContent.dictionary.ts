interface TaskTabsContentDictionary {
  NO_TASKS_TITLE: string;
  NO_TASKS_SUBTITLE: string;
}

const TASK_TABS_CONTENT_DICTIONARY: TaskTabsContentDictionary = {
  NO_TASKS_TITLE: 'No Tasks Found',
  NO_TASKS_SUBTITLE: 'You can always add a new task',
};

export default TASK_TABS_CONTENT_DICTIONARY;
