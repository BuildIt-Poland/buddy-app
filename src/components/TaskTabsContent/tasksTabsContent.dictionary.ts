interface TaskTabsContent {
  TITLE_UNCOMPLETED: string;
  TITLE_COMPLETED: string;
}

const DICTIONARY: TaskTabsContent = {
  TITLE_UNCOMPLETED: 'Todo',
  TITLE_COMPLETED: 'Done',
};

export default DICTIONARY;
