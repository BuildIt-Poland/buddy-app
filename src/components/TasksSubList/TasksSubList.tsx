import React from 'react';
import ListSubheader from '@material-ui/core/ListSubheader';
import TaskCheckbox from '../TaskCheckbox';
import { TasksSubListProps } from './types';

const TasksSubList: React.FC<TasksSubListProps> = ({
  tasks,
  count,
  title,
  onChange,
}) => {
  return (
    <>
      <ListSubheader>
        <strong>
          {title} ({count})
        </strong>
      </ListSubheader>
      {tasks.map(({ id, title, status }) => (
        <TaskCheckbox
          key={id}
          title={title}
          id={id}
          status={status}
          onChange={onChange}
        />
      ))}
    </>
  );
};

export default TasksSubList;
