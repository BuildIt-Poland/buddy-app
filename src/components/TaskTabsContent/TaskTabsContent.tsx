import React, { useMemo } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import List from '@material-ui/core/List';
import { TaskStatus } from 'buddy-app-schema';
import TaskListPlaceHolder from 'components/TaskListPlaceHolder';
import Typography from '@material-ui/core/Typography';
import Box from '@material-ui/core/Box';
import TasksSubList, { TasksSubListProps } from '../TasksSubList';
import { TaskTabsContentProps, TransformedTasks } from './types';
import DICTIONARY from './dictionary';

const useStyles = makeStyles(() => ({
  emptyContainer: {
    display: 'flex',
    flexDirection: 'column',
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    textAlign: 'center',
  },
}));

const transformTasks = (tasks: TaskTabsContentProps['tasks']): TransformedTasks => {
  const emptyTransformedTasks = {
    uncompletedTasks: [],
    completedTasks: [],
  } as TransformedTasks;

  if (tasks) {
    return tasks.reduce((result, task) => {
      if (task.status === TaskStatus.Completed) {
        result.completedTasks.push(task);
      } else {
        result.uncompletedTasks.push(task);
      }
      return result;
    }, emptyTransformedTasks);
  } else {
    return emptyTransformedTasks;
  }
};

const TaskTabsContent: React.FC<TaskTabsContentProps> = ({
  tasks,
  onChange,
  loading,
  tabIndex = 0,
}) => {
  const { emptyContainer } = useStyles();

  const { uncompletedTasks, completedTasks } = useMemo(() => transformTasks(tasks), [
    tasks,
  ]);
  const tasksList: TasksSubListProps[] = [
    {
      title: DICTIONARY.TITLE_UNCOMPLETED,
      tasks: uncompletedTasks,
    },
    {
      title: DICTIONARY.TITLE_COMPLETED,
      tasks: completedTasks,
    },
  ];
  const isEmptyList = uncompletedTasks.length === 0 && completedTasks.length === 0;

  const EmptyStateTaskList = () => (
    <Box className={emptyContainer}>
      <Typography variant={'h2'} component={'h2'}>
        {DICTIONARY.NO_TASKS_TITLE}
      </Typography>
    </Box>
  );

  return (
    <>
      {loading && <TaskListPlaceHolder />}
      {!loading && isEmptyList && <EmptyStateTaskList />}
      {!loading && !isEmptyList && (
        <List>
          {tasksList.map((props, key) => (
            <TasksSubList
              key={key}
              onChange={onChange}
              tabIndex={tabIndex}
              {...props}
            />
          ))}
        </List>
      )}
    </>
  );
};

export default TaskTabsContent;
