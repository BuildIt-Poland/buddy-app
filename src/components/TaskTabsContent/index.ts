import TaskTabsContent from './TaskTabsContent';

export default TaskTabsContent;
