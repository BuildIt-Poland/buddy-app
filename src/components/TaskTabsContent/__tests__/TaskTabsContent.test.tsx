import React from 'react';
import { create } from 'react-test-renderer';
import { tasksResponse } from '__mocks__';
import TasksSubList from 'components/TasksSubList';
import TaskTabsContent from '../';

jest.mock('@material-ui/core/Box', () => 'Box');
jest.mock('components/TasksSubList', () => 'TasksSubList');

describe('Component - TaskTabsContent', () => {
  test(`renders correctly`, () => {
    const component = create(
      <TaskTabsContent
        tasks={tasksResponse()}
        uncompletedCount={1}
        completedCount={1}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();

    //component.root.findAllByType(TasksSubList).length;
  });

  test(`renders correctly without tasks and counts`, () => {
    const component = create(<TaskTabsContent tasks={[]} />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
