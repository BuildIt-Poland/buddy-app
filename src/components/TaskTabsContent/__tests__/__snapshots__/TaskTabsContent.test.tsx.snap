// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - TaskTabsContent renders correctly 1`] = `
<Box
  className="makeStyles-wrapper-1"
>
  <TasksSubList
    count={1}
    tasks={
      Array [
        Object {
          "createdAt": "",
          "description": "",
          "id": "1",
          "implementationDate": "",
          "newbie": Object {},
          "status": "UNCOMPLETED",
          "title": "Call you head manager",
        },
      ]
    }
    title="Todo"
  />
  <TasksSubList
    count={1}
    tasks={
      Array [
        Object {
          "createdAt": "",
          "description": "",
          "id": "2",
          "implementationDate": "",
          "newbie": Object {},
          "status": "COMPLETED",
          "title": "Write you main manager",
        },
      ]
    }
    title="Done"
  />
</Box>
`;

exports[`Component - TaskTabsContent renders correctly without tasks and counts 1`] = `
<Box
  className="makeStyles-wrapper-1"
/>
`;
